### ISSUE TITLE
Tab does not focus next focusable element when elements are not in the light DOM

### ISSUE THREAD

**Issue by MartinMoizard** [PolymerElements/iron-overlay-behavior#188]
### Description
If the focusable elements are not in the light DOM of the iron-overlay-behavior component, tapping the tab key does not focus the next element.

### Expected outcome
The next element should be focused.

### Actual outcome
The focus is lost.

### Live Demo
https://jsfiddle.net/2q3fqehz/45/

### Steps to reproduce

- Clic in the 1st input
- Push the tab key on your keyboard
- The focus does not jump to the 2nd input

### Browsers Affected (tested on)
- [x] Chrome
- [x] Firefox
- [x] Safari 9

**Comment by MartinMoizard** [comment]
@valdrinkoshi reading the code, I would fix this issue by changing line 535 from
> if (nodeToCheck === nodeToSet) {
to
`if (nodeToCheck != null && nodeToSet != null && nodeToCheck === nodeToSet) {`
what do you think?

**Comment by valdrinkoshi** [comment]
Hi @MartinMoizard, overlays will search for first focusable and last focusable in the distributed children, while in your example you have focusable elements in the shadowDOM of another element.
From the POV of the overlay, it sees only `sample-content`, which is not focusable.
I'm working on creating `iron-focus-wrap-behavior` (see issue https://github.com/PolymerElements/iron-overlay-behavior/issues/179 and the PR https://github.com/PolymerElements/iron-overlay-behavior/pull/178). Most likely this behavior will live in a separate PR, so that it can be reused by other elements (e.g. your `sample-content`).

**Comment by valdrinkoshi** [comment]
Closing this as each component should handle focus within its light dom.

**Comment by wyattallen** [comment]
I'd like to second the suggestion @MartinMoizard makes for l.535. If the overlay fails to find the first and last focusable children, it checks whether `nodeToCheck === nodeToSet`, which passes because they are both undefined.

Moreover, @valdrinkoshi, you suggest that `each component should handle focus within its light dom`. How would you implement this in the sample @MartinMoizard provides? The code found in the JSFiddle definitely looks like it should work.

**Comment by valdrinkoshi** [comment]
From the POV of `iron-overlay-behavior`, there are no focusable nodes in the distributed nodes, hence it should prevent the TAB so that focus won't go out of the overlay (see [this comment](https://github.com/PolymerElements/iron-overlay-behavior/blob/master/iron-overlay-behavior.html#L535)). This should be done also in the case where there is only 1 focusable node.

For now, the best option is to distribute the focusable nodes directly in the `paper-dialog` http://jsbin.com/humufe/2/edit?html,output

I don't think we can fix this for now -- it's a current problem with focus and the shadow/shady dom, that could be fixed in the future when [`delegatesFocus`](https://github.com/w3c/webcomponents/issues/126) is implemented. 

Also, once and if `iron-focus-wrap-behavior` lands, `iron-overlay-behavior` will implement it (#178). You'll be able to set the first/last focusable of `paper-dialog` http://jsbin.com/tinafe/3/edit?html,output

**Comment by wyattallen** [comment]
Hi, @valdrinkoshi, thanks! This sounds like it should be an entry in the limitations section of the docs.

**Comment by valdrinkoshi** [comment]
Good idea @wibblymat! 
Another option is to add focusable placeholders before and after the `sample-content` in the `paper-dialog` http://jsbin.com/tinafe/7/edit?html,output
Like this, focus will wrap even if `sample-content` is not focusable.

**Comment by motss** [comment]
@valdrinkoshi Any solution for this example at the moment? http://jsbin.com/tojoroyibi/1/edit?html,output

**Comment by valdrinkoshi** [comment]
Hi @motss, unfortunately your example shows even more the current limitation: `iron-overlay-behavior` assumes its content will be distributed directly into the element, while in your case the content of `x-example` gets distributed into the `paper-dialog`.
So unfortunately no temporary solutions except moving the content inside the `paper-dialog`
```html
<paper-dialog>
  <p> content </p>
</paper-dialog>
```
The ideal solutions will be the one I talked about [in the previous comment](https://github.com/PolymerElements/iron-overlay-behavior/issues/188#issuecomment-223637262)
Hope it helps

**Comment by valdrinkoshi** [comment]
Looking into it a bit deeper, the problem is that `iron-overlay-behavior` searches into the children in the light dom without considering the distributed ones. I've opened a separate issue for it https://github.com/PolymerElements/iron-overlay-behavior/issues/194
